### DISCUSSION TITLE
Changes regarding the first superuser creation

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#5814] [Announcements]
> _Note: admins are renamed to superusers in v0.23.0-rc+ but they are the same thing._

Currently when there is no admin/superuser, PocketBase allows creating the first account by simply visiting `/_/` or by sending direct unauthenticated admins/superusers create API requests.

While it's convenient it can also lead to security issues depending on the type of app distribution in case developers forget to create the admin/superuser account when deploying their apps for the first time (ex. [Webhood#CVE-2024-31218](https://github.com/webhood-io/webhood/security/advisories/GHSA-h533-rxhm-73j2)).

**To prevent this from happening, with v0.23.0 I'm considering disabling the above behavior and remove the `RequireSuperuserAuthOnlyIfAny` middleware.**

Instead, on app start if we detect that there is no superuser account we'll:
- auto create a system superuser with a random secure password
- generate a non-refreshable short-lived (e.g. 30min) auth token for that system superuser
- print a URL in the terminal with the generated token that points to the UI installer where users can create their own superuser account or import a backup (the URL is auto opened in the browser via `xdg-open`/`rundll32`/`open` OR could be manually copy/pasted as fallback)

Once the custom superuser account is created, the system superuser is automatically deleted.

In the terminal, the first time starting PocketBase it will look something like:

```bash
[root@dev]$ ./pocketbase serve

2024/11/06 19:17:52 Server started at http://127.0.0.1:8090
├─ REST API:  http://127.0.0.1:8090/api/
└─ Dashboard: http://127.0.0.1:8090/_/

(!) Launch the URL below in the browser if it hasn't been open already to create your first superuser account:
http://127.0.0.1:8090/_/#/pbinstal/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJwYmNfMzE0MjYzNTgyMyIsImV4cCI6MTczMDkxNTI3MiwiaWQiOiI5NzU5NG5nYnU5MDJtZWQiLCJyZWZyZXNoYWJsZSI6ZmFsc2UsInR5cGUiOiJhdXRoIn0.30P1boc93wrG_Yg40vEx0lTfcEp2kKCfJ-6N7zXqsNY
(you can also create your first superuser account by running: ./pocketbase superuser upsert EMAIL PASS)
```

When running locally the above actually makes it easier than before because users no longer have to wonder what to do next BUT when PocketBase is started for the first time from a systemd service, deployed behind a reverse proxy (e.g. nginx) or in a container environment it will be required to do one of the following:

- create the first superuser programmatically either via migration or the `./pocketbase superuser create/upsert EMAIL PASS` command (_the `upsert` will overwrite existing superuser if it already exist making it suitable to always run as part of a CI/CD setup before starting the `serve` process_)

- copy/paste manually the URL from the stderr logs output, without the localhost part (this may not be very user friendly when deploying on platforms like Fly)

**I'll push later in the `develop` branch a POC implementation but I'm open for other suggestions if you think that it can be resolved better (especially the last bullet point).**

**Comment by gaganbiswas** [comment]
I think the cli command is the best go. You can do something similar to what medusa does and improve on it. Something like, `./pocketbase user create EMAIL PASS` will create a superuser (owner) if one doesn't exist and post that point will only create admins. The admins can't delete the owner or change the password. If the admins or owner forgets their password they can send a forgot password email to the registered address using a cli command like `./pocketbase user reset-password REGISTERED_EMAIL` or one can choose to create a admin user, login to the panel and send a forgot password request for the owner account.

**Reply by ganigeorgiev** [reply]
It is not "cli or nothing" type of situation. The `admin` (`superuser` in v0.23.0) command already exists and users can use it if they want to.

The problem is that when deploying for example behind a reverse proxy, users will no longer be able to just type `/_/` in the browser as before to setup their PocketBase instance when started for the first time and will have to do one of the suggested 2 options mentioned in last part of the post (_ideally I'm looking for suggestions how to avoid it and reduce the friction_).

**Reply by gaganbiswas** [reply]
The first point seems more logical in a reverse proxy or containerized environment.

For the second point we can do something like creating a secret during pocketbase setup _[`./pocketbase serve --secret=SUPERSECRET` or maybe in config or .env]_ which will be a one time use and will hold no significance once the superuser is created and then redirect the requests from `/_/SECRET` _(or some other URL)_ to `/#/pbinstal/eyJhbGciOi...`. Or `/_/` can have an input to enter the secret which will then redirect to the installer.

**Reply by ganigeorgiev** [reply]
I'm not sure how having a `/_/SECRET` or an input for user to enter a manual "secret" is much different and helps from what will be already printed with the process start.
It actually adds an extra step - the user to specify a manual "secret" with the serve command (that will be technically used just once) and I don't see it as improvement compared to just copying the `/_/#/pbinstall/TOKEN` part (_or running `./pocketbase superuser upsert EMAIL PASS` before the process start_).

**Reply by gaganbiswas** [reply]
Then, I think you already have given us an wonderful solution. 
- `./pocketbase superuser upsert EMAIL PASS` --> More user friendly during production than during local development
- `/_/#/pbinstall/TOKEN` --> More user friendly during local development than during production

**Reply by gaganbiswas** [reply]
> The first point seems more logical in a reverse proxy or containerized environment.
> 
> For the second point we can do something like creating a secret during pocketbase setup _[`./pocketbase serve --secret=SUPERSECRET` or maybe in config or .env]_ which will be a one time use and will hold no significance once the superuser is created and then redirect the requests from `/_/SECRET` _(or some other URL)_ to `/#/pbinstal/eyJhbGciOi...`. Or `/_/` can have an input to enter the secret which will then redirect to the installer.

I told this as on what to do if someone goes to `/_/` without create an initial superuser. You can alternatively just show the steps there on what to do next _[The two bullet points]_.

**Comment by ganigeorgiev** [comment]
I've pushed the POC in [v0.23.0-rc11](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.0-rc11). I'm still not entirely sure whether this is the best approach but it is definitely safer. Nothing is set in stone and I'm still open for other suggestions if someone have an idea or recommendation based on their experience with other similar apps.

**Comment by chnfyi** [comment]
Why did you prefer an URL with generated token ? Naively, i think that one possibility could be to print the auto created system superuser credentials instead and to prompt admins to authenticate on `/_/` immediatly to update the credentials, then :
- Let them updating credentials themselves
- Or force to update superuser credentials - _at least password_ - immediately upon first login

**Reply by ganigeorgiev** [reply]
Printing the password was also one of my first approach but decided against it because it will be visible in the stderr (and in the monitoring logs in various platforms like Fly) and also because configuring the superuser may not even be needed because the app could have its own separate interface and all other configurations could be applied via migrations (see the linked Webhood CVE above). Additionally printing the password will be possible only once but there is no guarantee that the user will init the application and on restart we'll remain with a system superuser for which we don't know the password (unless it is reseted on each startup but it just one more thing that we'll have to take care for).

**Reply by chnfyi** [reply]
Another approach (_maybe over-engineered, sorry_) could be to require MFA for first login and to print password (or to have a default password) and a short-lived OTP. Then prompt or force admin to update the superuser credentials.

**Reply by ganigeorgiev** [reply]
It is not possible to send an email OTP because the user would not have access to the dashboard to setup the SMTP credentials (the default `sendmail` command unfortunately is very unreliable).

**Reply by chnfyi** [reply]
I thought of printing the OTP in advance (based on a default email and a default password) instead of sending it by email, but again, i suggest it a little bit naively.

**Reply by ganigeorgiev** [reply]
Hm, I'm not sure. It will be less secure since OTPs are usually something very short and attacker can enumarete them easily if they know when you are deploying (which is very unlikely but still some security researchers will rise an eyebrow).

And at the end you still need to copy something from the terminal which is my main issue with the current approach - on local is fine due to the browser auto link open but for platforms like Fly, Railway, etc. printing stuff to the stderr is usually intercepted so that they can show the messages in their dashboard (_sometimes this also leads to problems related to the ordering of the messages if there too many at the same time_), which means that users will have to look for them among other messages related to their deployment.

**Reply by chnfyi** [reply]
Why not just keeping the auto created system superuser with `./pocketbase superuser upsert EMAIL PASS` as a unique option with an instruction like "_Create your first superuser account by running: ./pocketbase superuser upsert EMAIL PASS_" ?

**Reply by ganigeorgiev** [reply]
I think we are going in circles. The `superuser` command will have to be executed before starting the `serve` process. Additionally there are users that may not want to create a superuser because they just want to import a backup from the installer that already contains it (see https://github.com/pocketbase/pocketbase/issues/4534).

**Reply by chnfyi** [reply]
I meant, since url, password or OTPs are visible in the stderr, why not rely on `upsert` as a unique option (_still with auto created system superuser for security_) and add a step in introduction section from documentation ? I think it would remain user friendly. 

For example :

```
Once you've extracted the archive, in the same directory :

- Create your first superuser by running ./pocketbase superuser upsert EMAIL PASSWORD
- Then start directly the application by running ./pocketbase serve

And that's it! A web server will be started with the following routes: ...
```

Then users that want to import a backup would be able to do it directly from the UI after authenticate.

**Reply by ganigeorgiev** [reply]
But you miss the point where they'll have to specify an explicit email and secure password that may no longer be valid after the backup import.

The problem with the URL approach is also not that it is visible in the stderr but the fact that in some production deployments users users may need to search for that log and this may not always be easily done.

Additionally the `superuser upsert` command is already printed as suggestion when starting the executable below the url.

In any case I appreciate your suggestions but I think we are going in circles and I'll stop engage further.

**Reply by gaganbiswas** [reply]
> The problem with the URL approach is also not that it is visible in the stderr but the fact that in some production deployments users users may need to search for that log and this may not always be easily done.

If during the serve process the user creates a secret and then uses it with the url or maybe in the input field at `/_/`, i guess then that would be more easier than searching through the logs. Yes it will be an extra step, but its secure i guess, as well as one doesn't have to go through the logs. 

This can be optional as well. Like if someone doesn't create a secret they can simply get the url from the logs.

**Comment by tegarimansyah** [comment]
How about we also allow using env variable? For example in postgres docker compose

```
  db:
    image: postgres
    environment:
      POSTGRES_PASSWORD: example
```

It only works for first time when there is no superadmin. We can remove the password from env later and there will be no trace of password.

**Reply by ganigeorgiev** [reply]
It would require a prerequisite step and doesn't solve https://github.com/pocketbase/pocketbase/issues/4534. It also doesn't address the app distribution cases similar to the linked Webhood's CVE.

If users want to use env variables, they can do that already - either as part of the `superuser upsert` command, the `OnBootstrap` hook or a migration if they want to be a one-time operation only.

**Comment by markusleh** [comment]
As the developer responsible for [Webhood#CVE-2024-31218](https://github.com/webhood-io/webhood/security/advisories/GHSA-h533-rxhm-73j2) I think this is a good solution. 

My product does not use the superuser account for anything and I had improperly assumed that, as the admin user creation happens in `/_/` path, simply preventing access to that path via a reverse proxy would block all the superuser actions. 

Of course, in reality, this still leaves `/api/admins` endpoint exposed. I think a part of the cause of the vulnerability was also the fact that there exists a state where you can call `POST: /api/admins` without authorization (i.e. `RequireSuperuserAuthOnlyIfAny`) and this was not documented anywhere except in the source code. 

To be clear, the root cause for CVE-2024-31218 was not in Pocketbase. It has been clear to me from the start that, although possible, shipping Pocketbase as part of another product is not the main use case and a state where Pocketbase is ran on production with no superuser being created is likely only really an issue in my particular use case.

**Comment by huglemon** [comment]
I just saw this message, but I deployed it through dokploy, and the URL was not printed in the logs. Why is this?

**Reply by ganigeorgiev** [reply]
I have no experience with Dokploy and I cannot help you with it.

For why it hasn't printed anything - either you are not checking in the correct place for the logs or you are using an older PocketBase version. In all cases, as mentioned in your recently opened discussion https://github.com/pocketbase/pocketbase/discussions/6036, you can always create the superuser account manually via the `./pocketbase superuser EMAIL PASS` command (if you are using a custom `pb_data` you need to specify `--dir` too with the `superuser` command).

**Comment by hortinstein** [comment]
I **think** this broke one of the ways I was testing pocketbase, by bootstrapping a DB after launching it from docker.  I am using the 0.24.4 version of pocketbase from ```ghcr.io/muchobien/pocketbase:latest```:
 
```curl -X POST 'http://localhost:8091/api/admins/auth-with-password' -H 'Content-Type: application/json' -d '{"email":"me@automatedtests.com", "password":"automatedtests"}'```

I get the following: 

```{"data":{},"message":"The requested resource wasn't found.","status":404}``` what is the latest way to do this?

EDIT: additionally i did try the following init script:
```
#!/bin/sh

set -e  # Exit on any command failure

echo "Starting PocketBase initialization..."

# Debug: Print environment variables
echo "Admin Email: $POCKETBASE_ADMIN_EMAIL"
echo "Admin Password: $POCKETBASE_ADMIN_PASSWORD"

# Run the superuser upsert command
/usr/local/bin/pocketbase superuser upsert $POCKETBASE_ADMIN_EMAIL $POCKETBASE_ADMIN_PASSWORD

echo "Superuser upsert completed. Starting PocketBase server..."

# Start the PocketBase server
/usr/local/bin/pocketbase serve --http=0.0.0.0:8090 --dir=/pb_data --publicDir=/pb_public --hooksDir=/pb_hooks --dev
```

**Reply by ganigeorgiev** [reply]
@hortinstein I don't see how your issue is related to the above discussion.

It is not clear what actually is the problem you are trying to explain, nor what version you were using before and how you are testing.

Note that `/api/admins/...` is no longer available with PocketBase v0.23+. For more details please refer to the [v0.23+ changelog and its linked upgrade guide](https://github.com/pocketbase/pocketbase/blob/master/CHANGELOG.md#v0230).

Additionally, in the above provided init script you need to specify the same `--dir` flag for the `upsert` command, otherwise the executable will fallback to the default one.

If you have further questions, please open a new Q&A discussion but try to provide a little more context about the problem and your project to understand the bigger picture.

**Comment by neilyoung** [comment]
I don't know how many "superuser" creation requests I got today already, and satisfied. It is constantly again asking for it. A PITA, if you ask me

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean by this. The superuser creation should be asked only once. If not, then please create a minimal reproducible and clear steps to reproduce the issue, otherwise I'm not sure how to help you.

In any case, this is an old announcement that is already implemented. If you are experiencing issues, please open an issue or Q&A discussion.

**Comment by neilyoung** [comment]
Here is your sample code:

main.go:

```
package main

import "github.com/pocketbase/pocketbase"


func main() {

	app := pocketbase.New()

	app.Start()
}
```

go.mod:

```
module example.com/admin

go 1.24.4

require github.com/pocketbase/pocketbase v0.28.4

require (
	github.com/asaskevich/govalidator v0.0.0-20230301143203-a9d515a09cc2 // indirect
	github.com/disintegration/imaging v1.6.2 // indirect
	github.com/domodwyer/mailyak/v3 v3.6.2 // indirect
	github.com/dustin/go-humanize v1.0.1 // indirect
	github.com/fatih/color v1.18.0 // indirect
	github.com/gabriel-vasile/mimetype v1.4.9 // indirect
	github.com/ganigeorgiev/fexpr v0.5.0 // indirect
	github.com/go-ozzo/ozzo-validation/v4 v4.3.0 // indirect
	github.com/golang-jwt/jwt/v5 v5.2.2 // indirect
	github.com/google/uuid v1.6.0 // indirect
	github.com/inconshreveable/mousetrap v1.1.0 // indirect
	github.com/mattn/go-colorable v0.1.14 // indirect
	github.com/mattn/go-isatty v0.0.20 // indirect
	github.com/ncruces/go-strftime v0.1.9 // indirect
	github.com/pocketbase/dbx v1.11.0 // indirect
	github.com/remyoudompheng/bigfft v0.0.0-20230129092748-24d4a6f8daec // indirect
	github.com/spf13/cast v1.9.2 // indirect
	github.com/spf13/cobra v1.9.1 // indirect
	github.com/spf13/pflag v1.0.6 // indirect
	golang.org/x/crypto v0.39.0 // indirect
	golang.org/x/exp v0.0.0-20250606033433-dcc06ee1d476 // indirect
	golang.org/x/image v0.28.0 // indirect
	golang.org/x/net v0.41.0 // indirect
	golang.org/x/oauth2 v0.30.0 // indirect
	golang.org/x/sync v0.15.0 // indirect
	golang.org/x/sys v0.33.0 // indirect
	golang.org/x/text v0.26.0 // indirect
	modernc.org/libc v1.65.10 // indirect
	modernc.org/mathutil v1.7.1 // indirect
	modernc.org/memory v1.11.0 // indirect
	modernc.org/sqlite v1.38.0 // indirect
)
```

Run in like so 

```
go run .  serve
```

Wait for the browser to pop up, login with whatever credentials, stop the app, rerun it. Browser will popup again, asking you for su creds.

MacOS latest, MBP M4

**Reply by neilyoung** [reply]
Background: Apart of login in with email/password I didn't do anything at the first request. Maybe that, but I was under the assumption, that the su is created automatically from my login. If that is not the case - my bad.

**Reply by neilyoung** [reply]
```
(!) Launch the URL below in the browser if it hasn't been open already to create your first superuser account:
http://127.0.0.1:8090/_/#/pbinstal/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJwYmNfMzE0MjYzNTgyMyIsImV4cCI6MTc1MDc0MjM1OSwiaWQiOiJxaWt0NnZvbGJqaWZiY2UiLCJyZWZyZXNoYWJsZSI6ZmFsc2UsInR5cGUiOiJhdXRoIn0.MiM_B_3iWDyV21LBWcDj2PSoE_dPuC94PhIygRtFjBE
(you can also create your first superuser by running: /var/folders/7l/rc18f0m564qgtmjlzn6b_n5m0000gn/T/go-build735093424/b001/exe/admin superuser upsert EMAIL PASS)
```

```
/var/folders/7l/rc18f0m564qgtmjlzn6b_n5m0000gn/T/go-build735093424/b001/exe/admin superuser upsert admin@example.com 1111111111
bash: /var/folders/7l/rc18f0m564qgtmjlzn6b_n5m0000gn/T/go-build735093424/b001/exe/admin: No such file or directory
```

I'm sure it is me

**Reply by ganigeorgiev** [reply]
@neilyoung I've just tested it locally and I'm not able to reproduce it. I'm testing on Linux/Fedora and don't have access to a Mac but there is no platform specific code related to the superuser creation that could cause it to fail in the way you are explaining.

Your issue looks surprisingly similar to https://github.com/pocketbase/pocketbase/discussions/6920 but again, without clear steps to trigger the error I cannot assist you.

Similar to the other user, I suspect that there is something else in your environment or code (custom hooks, `pb_data` permissions, etc.) that you are not sharing.

If you think that it is a PocketBase bug, feel free to create a PR and I'll try to have a look at it but the automated and my local tests for now shows otherwise and I don't have the time to waste on this.

Fro clarity, here are the steps I've performed:

1. Create a main.go with the above content and initialized go mod.
2. Run `go run . serve`
3. A browser window popup for the superuser creation (there is also a log in the terminal with the superuser create url).
4. Create the super user.
5. Stop the application.
6. Start again.
7. No new superuser setup popup show up (the terminal also no longer prints the superuser create url).
8. Repeat again step 5 to double confirm.

**Reply by ganigeorgiev** [reply]
@neilyoung OK, I think I know what you are doing and why it doesn't work.

When you run with `go run`, Go builds your program in a temp location and starts it.

The output in the terminal is a little confusing and I'll think how to change it to make it more clear _(Done)_, but you shouldn't run manually that generated temp binary OR at least you should specify explicitly the location of your `pb_data` if you are going to do that with the `--dir` flag.

If you want to create the superuser from the console then you should either run first `go build` and starts that created executable or run something like:

```
go run . superuser upsert admin@example.com 1111111111
```

**Reply by ganigeorgiev** [reply]
In any case, please don't spam further as this is not the correct place for this type of posts. If you still need further help, please open a new Q&A discussion.
